File: xform-go/lexer.go
package xform

import (
	"fmt"
	"strings"
	"unicode"
)

type TokenKind string

const (
	TokEOF    TokenKind = "EOF"
	TokKW     TokenKind = "KW"
	TokIdent  TokenKind = "IDENT"
	TokOp     TokenKind = "OP"
	TokPunct  TokenKind = "PUNCT"
	TokString TokenKind = "STRING"
	TokNumber TokenKind = "NUMBER"
	TokDot    TokenKind = "DOT"
	TokSlash  TokenKind = "SLASH"
	TokAt     TokenKind = "AT"
)

type Token struct {
	Kind TokenKind
	Val  string
	Pos  int
}

var keywords = map[string]bool{
	"xform":   true,
	"version": true,
	"import":  true,
	"as":      true,
	"ns":      true,
	"def":     true,
	"var":     true,
	"let":     true,
	"in":      true,
	"for":     true,
	"where":   true,
	"return":  true,
	"if":      true,
	"then":    true,
	"else":    true,
	"match":   true,
	"case":    true,
	"default": true,
	"and":     true,
	"or":      true,
	"not":     true,
	"div":     true,
	"mod":     true,
	"rule":    true,
}

type Lexer struct {
	Text   string
	Pos    int
	Buffer *Token
}

func NewLexer(text string) *Lexer {
	return &Lexer{Text: text, Pos: 0, Buffer: nil}
}

func (l *Lexer) Peek() Token {
	if l.Buffer == nil {
		tok := l.nextToken()
		l.Buffer = &tok
	}
	return *l.Buffer
}

func (l *Lexer) Next() Token {
	if l.Buffer != nil {
		tok := *l.Buffer
		l.Buffer = nil
		return tok
	}
	return l.nextToken()
}

func (l *Lexer) Expect(kind TokenKind, value string) Token {
	tok := l.Next()
	if tok.Kind != kind || (value != "" && tok.Val != value) {
		panic(fmt.Errorf("expected %s %s at %d", kind, value, tok.Pos))
	}
	return tok
}

func (l *Lexer) ClearBuffer() {
	l.Buffer = nil
}

func (l *Lexer) skipWsComments() {
	for l.Pos < len(l.Text) {
		ch := l.Text[l.Pos]
		if unicode.IsSpace(rune(ch)) {
			l.Pos++
			continue
		}
		if ch == '#' {
			for l.Pos < len(l.Text) && l.Text[l.Pos] != '\n' {
				l.Pos++
			}
			continue
		}
		break
	}
}

func (l *Lexer) nextToken() Token {
	l.skipWsComments()
	if l.Pos >= len(l.Text) {
		return Token{Kind: TokEOF, Val: "", Pos: l.Pos}
	}

	ch := l.Text[l.Pos]

	if ch == ':' && l.Pos+1 < len(l.Text) && l.Text[l.Pos+1] == '=' {
		start := l.Pos
		l.Pos += 2
		return Token{Kind: TokOp, Val: ":=", Pos: start}
	}

	if strings.Contains("(){}[],:;", string(ch)) {
		l.Pos++
		return Token{Kind: TokPunct, Val: string(ch), Pos: l.Pos - 1}
	}

	if ch == '.' {
		start := l.Pos
		if l.Pos+1 < len(l.Text) && l.Text[l.Pos:l.Pos+2] == ".." {
			l.Pos += 2
			return Token{Kind: TokDot, Val: "..", Pos: start}
		}
		if l.Pos+2 < len(l.Text) && l.Text[l.Pos:l.Pos+3] == ".//" {
			l.Pos += 3
			return Token{Kind: TokDot, Val: ".//", Pos: start}
		}
		l.Pos++
		return Token{Kind: TokDot, Val: ".", Pos: start}
	}

	if ch == '/' {
		start := l.Pos
		if l.Pos+1 < len(l.Text) && l.Text[l.Pos:l.Pos+2] == "//" {
			l.Pos += 2
			return Token{Kind: TokSlash, Val: "//", Pos: start}
		}
		l.Pos++
		return Token{Kind: TokSlash, Val: "/", Pos: start}
	}

	if strings.Contains("<>=!+-*", string(ch)) {
		start := l.Pos
		l.Pos++
		if l.Pos < len(l.Text) && l.Text[l.Pos] == '=' {
			l.Pos++
			return Token{Kind: TokOp, Val: l.Text[start:l.Pos], Pos: start}
		}
		return Token{Kind: TokOp, Val: string(ch), Pos: start}
	}

	if ch == '\'' || ch == '"' {
		quote := ch
		start := l.Pos
		l.Pos++
		out := make([]rune, 0)
		for l.Pos < len(l.Text) {
			c := l.Text[l.Pos]
			if c == '\\' {
				l.Pos++
				if l.Pos >= len(l.Text) {
					break
				}
				esc := l.Text[l.Pos]
				switch esc {
				case 'n':
					out = append(out, '\n')
				case 't':
					out = append(out, '\t')
				case 'r':
					out = append(out, '\r')
				case 'u':
					if l.Pos+4 < len(l.Text) {
						hex := l.Text[l.Pos+1 : l.Pos+5]
						out = append(out, rune(parseHex(hex)))
						l.Pos += 4
					}
				default:
					out = append(out, rune(esc))
				}
				l.Pos++
				continue
			}
			if c == quote {
				l.Pos++
				return Token{Kind: TokString, Val: string(out), Pos: start}
			}
			out = append(out, rune(c))
			l.Pos++
		}
		panic(fmt.Errorf("unterminated string at %d", start))
	}

	if unicode.IsDigit(rune(ch)) {
		start := l.Pos
		for l.Pos < len(l.Text) {
			c := l.Text[l.Pos]
			if !unicode.IsDigit(rune(c)) && c != '.' {
				break
			}
			l.Pos++
		}
		return Token{Kind: TokNumber, Val: l.Text[start:l.Pos], Pos: start}
	}

	if unicode.IsLetter(rune(ch)) || ch == '_' {
		start := l.Pos
		for l.Pos < len(l.Text) {
			c := l.Text[l.Pos]
			if c == ':' {
				if l.Pos+1 < len(l.Text) && (unicode.IsLetter(rune(l.Text[l.Pos+1])) || unicode.IsDigit(rune(l.Text[l.Pos+1])) || l.Text[l.Pos+1] == '_' || l.Text[l.Pos+1] == '-') {
					l.Pos++
					continue
				}
				break
			}
			if !(unicode.IsLetter(rune(c)) || unicode.IsDigit(rune(c)) || c == '_' || c == '-') {
				break
			}
			l.Pos++
		}
		val := l.Text[start:l.Pos]
		if keywords[val] {
			return Token{Kind: TokKW, Val: val, Pos: start}
		}
		return Token{Kind: TokIdent, Val: val, Pos: start}
	}

	if ch == '@' {
		l.Pos++
		return Token{Kind: TokAt, Val: "@", Pos: l.Pos - 1}
	}

	panic(fmt.Errorf("unexpected character %q at %d", ch, l.Pos))
}

func parseHex(s string) int {
	val := 0
	for i := 0; i < len(s); i++ {
		val <<= 4
		c := s[i]
		switch {
		case c >= '0' && c <= '9':
			val += int(c - '0')
		case c >= 'a' && c <= 'f':
			val += int(c-'a') + 10
		case c >= 'A' && c <= 'F':
			val += int(c-'A') + 10
		}
	}
	return val
}

File: xform-ts/src/lexer.ts
export type TokenKind = "EOF" | "KW" | "IDENT" | "OP" | "PUNCT" | "STRING" | "NUMBER" | "DOT" | "SLASH" | "AT";

export interface Token {
  kind: TokenKind;
  value: string;
  pos: number;
}

const KEYWORDS = new Set([
  "xform",
  "version",
  "import",
  "as",
  "ns",
  "def",
  "var",
  "let",
  "in",
  "for",
  "where",
  "return",
  "if",
  "then",
  "else",
  "match",
  "case",
  "default",
  "and",
  "or",
  "not",
  "div",
  "mod",
  "rule",
]);

export class Lexer {
  private text: string;
  pos: number;
  private buffer: Token | null;

  constructor(text: string) {
    this.text = text;
    this.pos = 0;
    this.buffer = null;
  }

  peek(): Token {
    if (this.buffer === null) {
      this.buffer = this.nextToken();
    }
    return this.buffer;
  }

  next(): Token {
    if (this.buffer !== null) {
      const tok = this.buffer;
      this.buffer = null;
      return tok;
    }
    return this.nextToken();
  }

  expect(kind: TokenKind, value?: string): Token {
    const tok = this.next();
    if (tok.kind !== kind || (value !== undefined && tok.value !== value)) {
      throw new Error(`Expected ${kind} ${value ?? ""} at ${tok.pos}`);
    }
    return tok;
  }

  private skipWsComments(): void {
    while (this.pos < this.text.length) {
      const ch = this.text[this.pos];
      if (/\s/.test(ch)) {
        this.pos += 1;
        continue;
      }
      if (ch === "#") {
        while (this.pos < this.text.length && this.text[this.pos] !== "\n") {
          this.pos += 1;
        }
        continue;
      }
      break;
    }
  }

  private nextToken(): Token {
    this.skipWsComments();
    if (this.pos >= this.text.length) {
      return { kind: "EOF", value: "", pos: this.pos };
    }

    const ch = this.text[this.pos];

    if (ch === ":" && this.text[this.pos + 1] === "=") {
      const start = this.pos;
      this.pos += 2;
      return { kind: "OP", value: ":=", pos: start };
    }

    if ("(){}[],:;".includes(ch)) {
      this.pos += 1;
      return { kind: "PUNCT", value: ch, pos: this.pos - 1 };
    }

    if (ch === ".") {
      const start = this.pos;
      if (this.text.slice(this.pos, this.pos + 2) === "..") {
        this.pos += 2;
        return { kind: "DOT", value: "..", pos: start };
      }
      if (this.text.slice(this.pos, this.pos + 3) === ".//") {
        this.pos += 3;
        return { kind: "DOT", value: ".//", pos: start };
      }
      this.pos += 1;
      return { kind: "DOT", value: ".", pos: start };
    }

    if (ch === "/") {
      const start = this.pos;
      if (this.text.slice(this.pos, this.pos + 2) === "//") {
        this.pos += 2;
        return { kind: "SLASH", value: "//", pos: start };
      }
      this.pos += 1;
      return { kind: "SLASH", value: "/", pos: start };
    }

    if ("<>=!+-*".includes(ch)) {
      const start = this.pos;
      this.pos += 1;
      if (this.text[this.pos] === "=") {
        this.pos += 1;
        return { kind: "OP", value: this.text.slice(start, this.pos), pos: start };
      }
      return { kind: "OP", value: ch, pos: start };
    }

    if (ch === "'" || ch === "\"") {
      const quote = ch;
      const start = this.pos;
      this.pos += 1;
      const out: string[] = [];
      while (this.pos < this.text.length) {
        const c = this.text[this.pos];
        if (c === "\\") {
          this.pos += 1;
          if (this.pos >= this.text.length) break;
          const esc = this.text[this.pos];
          if (esc === "n") out.push("\n");
          else if (esc === "t") out.push("\t");
          else if (esc === "r") out.push("\r");
          else if (esc === "u") {
            const hex = this.text.slice(this.pos + 1, this.pos + 5);
            out.push(String.fromCharCode(parseInt(hex, 16)));
            this.pos += 4;
          } else out.push(esc);
          this.pos += 1;
          continue;
        }
        if (c === quote) {
          this.pos += 1;
          return { kind: "STRING", value: out.join(""), pos: start };
        }
        out.push(c);
        this.pos += 1;
      }
      throw new Error(`Unterminated string at ${start}`);
    }

    if (/\d/.test(ch)) {
      const start = this.pos;
      while (this.pos < this.text.length && /[\d.]/.test(this.text[this.pos])) {
        this.pos += 1;
      }
      return { kind: "NUMBER", value: this.text.slice(start, this.pos), pos: start };
    }

    if (/[A-Za-z_]/.test(ch)) {
      const start = this.pos;
      while (this.pos < this.text.length) {
        const c = this.text[this.pos];
        if (c === ":") {
          const next = this.text[this.pos + 1];
          if (next && /[A-Za-z0-9_-]/.test(next)) {
            this.pos += 1;
            continue;
          }
          break;
        }
        if (!/[A-Za-z0-9_-]/.test(c)) break;
        this.pos += 1;
      }
      const val = this.text.slice(start, this.pos);
      if (KEYWORDS.has(val)) {
        return { kind: "KW", value: val, pos: start };
      }
      return { kind: "IDENT", value: val, pos: start };
    }

    if (ch === "@") {
      this.pos += 1;
      return { kind: "AT", value: "@", pos: this.pos - 1 };
    }

    throw new Error(`Unexpected character ${JSON.stringify(ch)} at ${this.pos}`);
  }

  clearBuffer(): void {
    this.buffer = null;
  }
}
